aodntools/timeseries_products/gridded_timeseries.py

```diff
@@ -0,0 +1,260 @@
+import numpy as np
+import bisect
+import argparse
+import os.path
+import json
+from datetime import datetime
+
+import xarray as xr
+import pandas as pd
+
+import aggregated_timeseries as TStools
+
+
+TEMPLATE_JSON = 'gridded_timeseries_template.json'
+
+
+def make_depth_bins(nc, increment=10):
+    """
+    generate a list of depth bins from min/max DEPTH.
+    Re-scale the min/max to nearest increment round number and start the bin from there
+    :param nc: hourly dataset
+    :return: list of depth bins
+    """
+
+    dmin = int(nc.DEPTH.min() / increment) * increment
+    dmax = int(nc.DEPTH.max() / increment) * increment + increment
+    dbin = list(range(dmin, dmax, increment))
+
+    return dbin
+
+def adjust_depth_bins(depth_bins, depth_min, depth_max):
+    """
+    Adjust the provided depth bins to match the min and max registered depth
+    :param depth_bins: list of target depths
+    :param depth_min: min valid depth
+    :param depth_max: max valid depth
+    :return: list of adjusted target depths
+    """
+
+    target_depth_min = bisect.bisect_left(depth_bins, depth_min)
+    target_depth_max = bisect.bisect_right(depth_bins, depth_max)
+    return depth_bins[target_depth_min:target_depth_max]
+
+
+def get_depth_mask(depth_bins, depths, max_separation):
+    """
+    return a boolean mask of depth bins where the interpolation is possible
+    based on maximum separation between the instruments
+    :param depth_bins: list of depth bins where the interpolation will happen
+    :param depths: list of depth of the instruments at the current timestamp
+    :param max_separation: maximum allowable separation between readings
+    :return: list of depth masked
+    """
+    depth_mask = [False] * len(depth_bins)
+    for depth_index in range(len(depths) - 1):
+        if depths[depth_index + 1] - depths[depth_index] < max_separation:
+            depth_min = bisect.bisect_left(depth_bins, depths[depth_index])
+            depth_max = bisect.bisect_right(depth_bins, depths[depth_index + 1])
+            depth_interval_len = depth_max - depth_min
+            depth_mask[depth_min:depth_max] = [True] * depth_interval_len
+    return depth_mask
+
+def sort_attributes(attributes):
+    """
+
+    :param attributes: dictionary of global attributes
+    :return: sorted dictionary
+    """
+    return dict(sorted(attributes.items()))
+
+def sort_depths(depths, values):
```

Comment:
You are mutating the arguments, this is not recommended.

Also, I don't think your argument types (lists) are the best. In this case arrays are much better since you are doing index operations.

If you use numpy and are happy in using arrays, a much cleaner version is possible:

```
sorted_indexes = np.argsort(depths)
sorted_depths = depths[sorted_indexes]
sorted_values = values[sorted_indexes]
```

Reply:
fixed mutation of the arguments. 
Using numpy array is slower...

Reply:
The type makes a difference:

```
def sort_values_by_depth(depth,values)
       x=np.argsort(depth)
       return depth[x],values[x]
```

The above take 20us here.
The list code takes 130-150us and it's also longer and more verbose.

If I need to convert types, then the above code will take 170-190us, not much more for a shorter and less verbose code:

```
def sort_values_by_depth(depth,values)
       x=np.argsort(depth)
       return np.array(depth)[x], np.array(values)[x]
```

Reply:
I agree that the numpy approach is cleaner but I cannot replicate your results. My tests showed that the list-approach sorting is almost 3 times faster that the np.array approach in small lists. Longer the list, smaller the difference, but we're not expecting to process lists longer that ~100 elements:

list approach:
```
depth = list(range(0,100, 10))+[15, 98, 43]
values = list(range(len(depth)))

In [10]: %timeit sort_depths(depth, values)
2.45 µs ± 48.1 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```
np.array approach:
```
In [13]: %timeit sort_values_by_depth(depth, values)
7.07 µs ± 59.1 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

```

Reply:
as discussed, the speed is secondary here since the difference is not O(1) and the call isn't the bottleneck (which should be all the calls within the for loop - particularly the np.interp one).